### PR TITLE
Wip/hughsie/no capsule coalesce2

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [X] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupdplugin/fu-plugin.c [hunk 1]
```diff
@@ -507,6 +518,10 @@ fu_plugin_device_add (FuPlugin *self, FuDevice *device)
 		return;
 	}
 
+	/* add to array */
+	fu_plugin_ensure_devices (self);
+	g_ptr_array_add (priv->devices, g_object_ref (device));
```
Comment (superm1) [inline libfwupdplugin/fu-plugin.c:523]: are you missing the matching unref on remove?

Comment (hughsie) [reply libfwupdplugin/fu-plugin.c:523]: No, `g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);` speficies the function to use when deallocating or removing items from the array.

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c [hunk 2]
```diff
@@ -762,6 +762,38 @@ fu_plugin_unlock (FuPlugin *plugin, FuDevice *device, GError **error)
...
+			       fu_device_get_name (device),
+			       fu_device_get_id (device));
+	devices = fu_plugin_get_devices (plugin);
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device_tmp = g_ptr_array_index (devices, i);
+		if (device_tmp == device)
+			continue;
+		fu_device_remove_flag (device_tmp, FWUPD_DEVICE_FLAG_UPDATABLE);
+		fu_device_add_flag (device_tmp, FWUPD_DEVICE_FLAG_UPDATABLE_HIDDEN);
+		fu_device_set_update_message (device_tmp, msg);
```
Comment (superm1) [inline plugins/uefi-capsule/fu-plugin-uefi-capsule.c:793]: shouldn't this be update error?

**Review by superm1 — COMMENTED** [review]
Are there any plugins left tracking their own device cache in a `FuPluginData`?  If so those should probably be removed as a result of this change.

**Comment by superm1** [comment]
I think this should be documented in `uefi-capsule/README.md` as it's a pretty big difference between Lenovo's implementation and everyone elses'.

File: plugins/uefi-capsule/fu-plugin-uefi-capsule.c (hunk 2 above)
Comment (hughsie) [reply plugins/uefi-capsule/fu-plugin-uefi-capsule.c:793]: Hmm, I wasn't sure. It's not like the error was attempted and it failed, which I think is how we use update_error now.

Comment (superm1) [reply plugins/uefi-capsule/fu-plugin-uefi-capsule.c:793]: Update message is shown when an update is done.  It's stuff like the WD19 does where it lets you know the update finishes next time you unplug.  I think of Update Error as any time it's not updatable, or failed to update here is why.

Comment (hughsie) [reply plugins/uefi-capsule/fu-plugin-uefi-capsule.c:793]: Excellent, agreed -- fixed.

**Comment by hughsie** [comment]
> Are there any plugins left tracking their own device cache in a FuPluginData

Good question. I think the only plugin doing this is synaptics_mst, and I didn't convert this as it's not strictly the same thing. fu_plugin_get_devices() is getting all the *added* devices, and synaptics_mst is using the device array for devices that might exist right now or might appear in the future during a rescan -- which is subtly different.

**Comment by hughsie** [comment]
>  think this should be documented in uefi-capsule/README.md

Something like this perhaps?

	On some Lenovo hardware only one capsule is installable due to possible problems
	with the UpdateCapsule coalesce operation. As soon as one UEFI device has been
	scheduled for update the other UEFI devices found in the ESRT will be marked
	as `updatable-hidden` rather than `updatable`. Rebooting will restore them so
	they can be updated on next OS boot.

**Comment by superm1** [comment]
I don't know what that HWID corresponds to, is it just some Lenovo hardware or all Lenovo hardware?  The text looks good otherwise for the capsule case.  Maybe also document the reason for the "Linux-Firmware-Updater" vs "Linux Firmware Updater" at same time though.

**Comment by hughsie** [comment]
> is it just some Lenovo hardware or all Lenovo hardware

At the moment it's all Lenovo hardware, which is suboptimal as we only really want to match on specific thinkpad models

>  Maybe also document the reason for the "Linux-Firmware-Updater

Good idea, done.

**Review by superm1 — APPROVED** [review]

**Pull request merged**